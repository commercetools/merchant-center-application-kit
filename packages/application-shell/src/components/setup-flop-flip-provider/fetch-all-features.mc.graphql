query AllFeatures {
  allFeatures {
    name
    value
    reason
  }
}
