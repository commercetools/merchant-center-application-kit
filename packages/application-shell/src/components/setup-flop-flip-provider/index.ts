export { default } from './setup-flop-flip-provider';
