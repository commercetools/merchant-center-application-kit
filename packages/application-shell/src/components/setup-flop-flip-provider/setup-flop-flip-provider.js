import React from 'react';
import PropTypes from 'prop-types';
import { defaultMemoize } from 'reselect';
import ldAdapter from '@flopflip/launchdarkly-adapter';
import { ConfigureFlopFlip } from '@flopflip/react-broadcast';

// This value is hard-coded here because we want to make sure that the
// app uses our account of LD. The value is meant to be public, so there
// is no need to be concerned about security.
const ldClientSideIdProduction = '5979d95f6040390cd07b5e01';
// On our staging env we use a different ID, therefore we need to use the
// one above only for `production` environment.
const ldClientSideIdStaging = '5979d95f6040390cd07b5e00';

export const getFlopflipReconfiguration = defaultMemoize(projectKey => ({
  custom: { project: projectKey },
}));

export class SetupFlopFlipProvider extends React.PureComponent {
  static displayName = 'SetupFlopFlipProvider';
  static propTypes = {
    user: PropTypes.shape({
      id: PropTypes.string.isRequired,
      launchdarklyTrackingId: PropTypes.string.isRequired,
      launchdarklyTrackingGroup: PropTypes.string.isRequired,
      launchdarklyTrackingTeam: PropTypes.array.isRequired,
      launchdarklyTrackingTenant: PropTypes.string.isRequired,
    }),
    defaultFlags: PropTypes.object,
    appEnv: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
  };

  createLaunchdarklyAdapterArgs = defaultMemoize(
    (
      clientSideId,
      userId,
      ldTrackingId,
      ldTrackingGroup,
      ldTrackingTeam,
      ldTrackingTenant
    ) => ({
      clientSideId,
      user: {
        key: userId,
        custom: {
          id: ldTrackingId,
          team: ldTrackingTeam,
          group: ldTrackingGroup,
          tenant: ldTrackingTenant,
        },
      },
    })
  );

  render() {
    return (
      <ConfigureFlopFlip
        adapter={ldAdapter}
        adapterArgs={this.createLaunchdarklyAdapterArgs(
          this.props.appEnv === 'production'
            ? ldClientSideIdProduction
            : ldClientSideIdStaging,
          this.props.user && this.props.user.id,
          this.props.user && this.props.user.launchdarklyTrackingId,
          this.props.user && this.props.user.launchdarklyTrackingGroup,
          this.props.user && this.props.user.launchdarklyTrackingTeam,
          this.props.user && this.props.user.launchdarklyTrackingTenant
        )}
        defaultFlags={this.props.defaultFlags}
        shouldDeferAdapterConfiguration={!this.props.user}
      >
        {this.props.children}
      </ConfigureFlopFlip>
    );
  }
}

export default SetupFlopFlipProvider;
