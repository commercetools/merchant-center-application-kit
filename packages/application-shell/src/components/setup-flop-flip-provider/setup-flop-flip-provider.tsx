import type { TFlags } from '@flopflip/types';
import type {
  TAllFeaturesQuery,
  TFetchLoggedInUserQuery,
} from '../../types/generated/mc';

import React from 'react';
import { useApplicationContext } from '@commercetools-frontend/application-shell-connectors';
import ldAdapter from '@flopflip/launchdarkly-adapter';
import httpAdapter from '@flopflip/http-adapter';
import combineAdapters from '@flopflip/combine-adapters';
import { ConfigureFlopFlip } from '@flopflip/react-broadcast';
import { useApolloClient } from '@apollo/client/react';
import { GRAPHQL_TARGETS } from '@commercetools-frontend/constants';
import useAllMenuFeatureToggles from '../../hooks/use-all-menu-feature-toggles';
import { FLAGS } from '../../feature-toggles';
import AllFeaturesQuery from './fetch-all-features.mc.graphql';

type Props = {
  projectKey?: string;
  user?: TFetchLoggedInUserQuery['user'];
  flags?: TFlags;
  defaultFlags?: TFlags;
  ldClientSideId?: string;
  children: React.ReactNode;
  shouldDeferAdapterConfiguration?: boolean;
};

type TLaunchDarklyUserCustomFields = {
  project: string;
  id: string;
  team: string[];
  group: string;
  subgroup: string;
  tenant: string;
};

type TFetchedHttpAdapterFlag = {
  name: string;
  value: boolean;
  reason?: string;
};
type TParsedHttpAdapterFlags = Record<
  string,
  {
    value: boolean;
    reason?: string;
  }
>;

// This value is hard-coded here because we want to make sure that the
// app uses our account of LD. The value is meant to be public, so there
// is no need to be concerned about security.
const ldClientSideIdProduction = '5979d95f6040390cd07b5e01';

function getUserCustomFieldsForLaunchDarklyAdapter(
  user?: Props['user'],
  projectKey?: string
): TLaunchDarklyUserCustomFields {
  return {
    project: projectKey ?? '',
    id: user?.launchdarklyTrackingId ?? '',
    team: user?.launchdarklyTrackingTeam ?? [],
    group: user?.launchdarklyTrackingGroup ?? '',
    subgroup: user?.launchdarklyTrackingSubgroup ?? '',
    tenant: user?.launchdarklyTrackingTenant ?? '',
  };
}

type TFetchedFlags = {
  allFeatures: TFetchedHttpAdapterFlag[];
};
const parseFlags = (fetchedFlags: TFetchedFlags): TParsedHttpAdapterFlags =>
  Object.fromEntries(
    fetchedFlags.allFeatures.map((fetchedFlag) => [
      fetchedFlag.name,
      {
        value: fetchedFlag.value,
        reason: fetchedFlag.reason,
      },
    ])
  );

export const SetupFlopFlipProvider = (props: Props) => {
  const apolloClient = useApolloClient();
  const enableFeatureConfigurationFetching = useApplicationContext(
    (context) => context.environment.enableFeatureConfigurationFetching
  );
  const allMenuFeatureToggles = useAllMenuFeatureToggles();
  const flags = React.useMemo(
    () => ({
      ...FLAGS,
      ...allMenuFeatureToggles.allFeatureToggles,
      ...props.flags,
    }),
    [allMenuFeatureToggles.allFeatureToggles, props.flags]
  );
  React.useMemo(() => {
    if (enableFeatureConfigurationFetching) {
      combineAdapters.combine([ldAdapter, httpAdapter]);
    } else {
      combineAdapters.combine([ldAdapter]);
    }
  }, [enableFeatureConfigurationFetching]);

  const defaultFlags = React.useMemo(
    () => ({
      ...FLAGS,
      ...allMenuFeatureToggles.allFeatureToggles,
      ...props.defaultFlags,
    }),
    [allMenuFeatureToggles.allFeatureToggles, props.defaultFlags]
  );

  const adapterArgs = React.useMemo(
    () => ({
      user: {
        key: props.user?.id,
      },
      launchdarkly: {
        sdk: {
          // Allow to overwrite the client ID, passed via the `additionalEnv` properties
          // of the application config.
          // This is mostly useful for internal usage on our staging environments.
          clientSideId: props.ldClientSideId ?? ldClientSideIdProduction,
        },
        flags,
        user: {
          key: props.user?.id,
          custom: getUserCustomFieldsForLaunchDarklyAdapter(
            props.user,
            props.projectKey
          ),
        },
      },
      http: {
        user: {
          key: props.user?.id,
        },
        execute: async function () {
          const response = await apolloClient.query<TAllFeaturesQuery>({
            query: AllFeaturesQuery,
            errorPolicy: 'ignore',
            context: {
              target: GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND,
            },
          });

          return parseFlags(response.data);
        },
      },
    }),
    [apolloClient, flags, props.ldClientSideId, props.projectKey, props.user]
  );

  /**
   * The `<ApplicationShell />` tests itself. When it does so it can not setup and use
   * the `launchdarkly-adapter` and `http-adapter`. As a result the `process.env`
   * is evaluated and a test provider is rendered for feature flags.
   *
   * The `process.env` check also ensures that this block is excluded from production
   * bundles.
   */
  if (process.env.NODE_ENV === 'test') {
    const { TestProviderFlopFlip } = require('@flopflip/react-broadcast');

    return (
      <TestProviderFlopFlip flags={defaultFlags}>
        {props.children}
      </TestProviderFlopFlip>
    );
  }

  return (
    <ConfigureFlopFlip<typeof combineAdapters>
      adapter={combineAdapters}
      adapterArgs={adapterArgs}
      defaultFlags={defaultFlags}
      shouldDeferAdapterConfiguration={
        typeof props.shouldDeferAdapterConfiguration === 'boolean'
          ? props.shouldDeferAdapterConfiguration
          : !props.user || allMenuFeatureToggles.isLoading
      }
    >
      {props.children}
    </ConfigureFlopFlip>
  );
};

SetupFlopFlipProvider.displayName = 'SetupFlopFlipProvider';

export default SetupFlopFlipProvider;
