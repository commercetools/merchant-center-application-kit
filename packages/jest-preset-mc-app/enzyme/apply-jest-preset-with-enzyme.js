const path = require('path');

const resolveRelativePath = (relativePath) =>
  path.resolve(__dirname, relativePath);

const applyJestPresetWithEnzyme = ({ enzymeAdapterVersion, jestPreset }) => {
  switch (enzymeAdapterVersion) {
    case 16:
    case 17:
      break;
    default:
      throw new Error(
        `Invalid Enzyme adapter version "${enzymeAdapterVersion}". Supported versions are "16", "17".`
      );
  }
  return {
    ...jestPreset,
    // Custom preset config specific for Enzyme.
    snapshotSerializers: [
      ...(jestPreset.snapshotSerializers || []),
      'enzyme-to-json/serializer',
    ],
    setupFilesAfterEnv: [
      ...(jestPreset.setupFilesAfterEnv || []),
      resolveRelativePath(
        `./setup-test-framework-for-enzyme-${enzymeAdapterVersion}.js`
      ),
    ],
  };
};

module.exports = applyJestPresetWithEnzyme;
