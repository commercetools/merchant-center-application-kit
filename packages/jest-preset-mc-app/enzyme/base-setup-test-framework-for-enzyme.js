require('jest-enzyme');

const configureEnzymeExtensions = require('@commercetools/enzyme-extensions');
const commerceToolsEnzymeMatchers = require('@commercetools/jest-enzyme-matchers');
const ShallowWrapper = require('enzyme/ShallowWrapper');

expect.extend({
  toBeComponentWithName(received, actual) {
    const pass = received && received.displayName === actual;
    const message = () =>
      `expected ${received} ${pass ? 'not ' : ''} to be ${actual}`;
    return { message, pass };
  },
});

expect.extend(commerceToolsEnzymeMatchers);
configureEnzymeExtensions(ShallowWrapper);
