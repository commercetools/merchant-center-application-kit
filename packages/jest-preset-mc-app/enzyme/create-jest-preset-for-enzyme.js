const path = require('path');
const defaultPreset = require('../jest-preset');

const resolveRelativePath = (relativePath) =>
  path.resolve(__dirname, relativePath);

const createJestPresetForEnzyme = (version) => ({
  ...defaultPreset,
  snapshotSerializers: ['enzyme-to-json/serializer'],
  setupFilesAfterEnv: [
    resolveRelativePath('../setup-test-framework.js'),
    resolveRelativePath(`./setup-test-framework-for-enzyme-${version}.js`),
  ],
});

module.exports = createJestPresetForEnzyme;
