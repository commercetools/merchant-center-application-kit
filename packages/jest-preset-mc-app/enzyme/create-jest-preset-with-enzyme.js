const path = require('path');

const resolveRelativePath = (relativePath) =>
  path.resolve(__dirname, relativePath);

const createJestPresetForEnzyme = ({ enzymeAdapterVersion, jestPreset }) => {
  switch (enzymeAdapterVersion) {
    case 16:
    case 17:
      break;
    default:
      throw new Error(
        `Invalid Enzyme adapter version "${enzymeAdapterVersion}". Supported versions are "16", "17".`
      );
  }
  return {
    // Use default preset
    ...jestPreset,
    snapshotSerializers: [
      ...(jestPreset.snapshotSerializers || []),
      'enzyme-to-json/serializer',
    ],
    setupFilesAfterEnv: [
      ...(jestPreset.setupFilesAfterEnv || []),
      resolveRelativePath(
        `./setup-test-framework-for-enzyme-${enzymeAdapterVersion}.js`
      ),
    ],
  };
};

module.exports = createJestPresetForEnzyme;
