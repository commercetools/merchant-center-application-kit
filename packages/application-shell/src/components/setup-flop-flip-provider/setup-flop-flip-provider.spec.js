import React from 'react';
import { shallow } from 'enzyme';
import { ConfigureFlopFlip } from '@flopflip/react-broadcast';
import { SetupFlopFlipProvider } from './setup-flop-flip-provider';

const createTestProps = props => ({
  user: {
    id: 'u1',
    launchdarklyTrackingId: '123',
    launchdarklyTrackingGroup: 'ct',
    launchdarklyTrackingTeam: ['abc', 'def'],
    launchdarklyTrackingTenant: 'xy',
  },
  defaultFlags: {
    flagA: true,
  },
  appEnv: 'staging',
  children: <div />,
  ...props,
});

describe('rendering', () => {
  let props;
  let wrapper;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<SetupFlopFlipProvider {...props} />);
  });
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should render <ConfigureFlopFlip>', () => {
    expect(wrapper).toRender(ConfigureFlopFlip);
  });
  it('should pass staging "clientSideId" as adapter arg', () => {
    expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
      'adapterArgs',
      expect.objectContaining({
        clientSideId: '5979d95f6040390cd07b5e00',
      })
    );
  });
  it('should pass "user.key" as adapter arg', () => {
    expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
      'adapterArgs',
      expect.objectContaining({
        user: expect.objectContaining({
          key: props.user.id,
        }),
      })
    );
  });
  it('should pass "user.custom.id" as adapter arg', () => {
    expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
      'adapterArgs',
      expect.objectContaining({
        user: expect.objectContaining({
          custom: expect.objectContaining({
            id: props.user.launchdarklyTrackingId,
          }),
        }),
      })
    );
  });
  it('should pass "user.custom.group" as adapter arg', () => {
    expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
      'adapterArgs',
      expect.objectContaining({
        user: expect.objectContaining({
          custom: expect.objectContaining({
            group: props.user.launchdarklyTrackingGroup,
          }),
        }),
      })
    );
  });
  it('should pass "user.custom.team" as adapter arg', () => {
    expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
      'adapterArgs',
      expect.objectContaining({
        user: expect.objectContaining({
          custom: expect.objectContaining({
            team: props.user.launchdarklyTrackingTeam,
          }),
        }),
      })
    );
  });
  it('should pass "user.custom.tenant" as adapter arg', () => {
    expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
      'adapterArgs',
      expect.objectContaining({
        user: expect.objectContaining({
          custom: expect.objectContaining({
            tenant: props.user.launchdarklyTrackingTenant,
          }),
        }),
      })
    );
  });
  describe('when user is not defined', () => {
    beforeEach(() => {
      props = createTestProps({ user: null });
      wrapper = shallow(<SetupFlopFlipProvider {...props} />);
    });
    it('should set "shouldDeferAdapterConfiguration" to true', () => {
      expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
        'shouldDeferAdapterConfiguration',
        true
      );
    });
  });
  describe('when user is defined', () => {
    it('should set "shouldDeferAdapterConfiguration" to false', () => {
      expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
        'shouldDeferAdapterConfiguration',
        false
      );
    });
  });
  describe('when env is production', () => {
    beforeEach(() => {
      wrapper = shallow(
        <SetupFlopFlipProvider {...props} appEnv="production" />
      );
    });
    it('should pass production "clientSideId" as adapter arg', () => {
      expect(wrapper.find(ConfigureFlopFlip)).toHaveProp(
        'adapterArgs',
        expect.objectContaining({
          clientSideId: '5979d95f6040390cd07b5e01',
        })
      );
    });
  });
});
