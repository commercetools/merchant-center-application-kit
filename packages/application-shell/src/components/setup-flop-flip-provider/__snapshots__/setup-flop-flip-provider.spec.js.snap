// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should match snapshot 1`] = `
<ConfigureFlopflip
  adapter={
    Object {
      "configure": [Function],
      "getFlag": [Function],
      "getIsReady": [Function],
      "reconfigure": [Function],
      "updateUserContext": [Function],
    }
  }
  adapterArgs={
    Object {
      "clientSideId": "5979d95f6040390cd07b5e00",
      "user": Object {
        "custom": Object {
          "group": "ct",
          "id": "123",
          "project": "test-project-key",
          "team": Array [
            "abc",
            "def",
          ],
          "tenant": "xy",
        },
        "key": "u1",
      },
    }
  }
  defaultFlags={
    Object {
      "flagA": true,
    }
  }
  shouldDeferAdapterConfiguration={false}
>
  <div />
</ConfigureFlopflip>
`;
